db/size_estimates_virtual_reader.cc

```diff
@@ -208,10 +208,10 @@ static future<std::vector<token_range>> get_local_ranges(database& db) {
         // All queries will be on that table, where all entries are text and there's no notion of
         // token ranges form the CQL point of view.
         auto left_inf = boost::find_if(ranges, [] (auto&& r) {
-            return !r.start() || r.start()->value() == dht::minimum_token();
+            return r.end() && (!r.start() || r.start()->value() == dht::minimum_token());
```

Comment:
I don't understand what this logic does - if r.end() then immediately we get left_inf true? what does it mean? Can't we have a finite range with both start and end?

Reply:
We don't - it's a necessary condition, because we later call `left_inf.end()`. We don't short-circuit here on true, since it's an `&&`, on the contrary - if `end()` is not present, then we immediately return false, because it's not a range with bounded end.

Reply:
Hmm, I guess with your definition, "left_inf" means the left, and *only* the left, is infinite (and the right is finite). I have no idea if that was the intended definition, though. I hope somebody who understand what this code was supposed to do (Kamil) can review it - I'm a little scared to commit this without understanding what this fixes...

Reply:
The range on which this code failed is `start == minimum_token, end == null` - it was recognized as a range "bounded from the left side, but bounded from the right side" - since these are the assumptions we rely on in the later code, and it's not the case. This range is not bounded, and calling `->end()` on it results in an assertion failure.

Reply:
**Unless** the ranges I presented in the counter-argument are for some reason illegal - if they are, then the actual root cause is somewhere else, but I think there's nothing inherently bad with these ranges, it's just `[minimum_token, inf)`.

Reply:
Ok, makes sense. I'll commit this quickly so the tests don't fail. I hope it's not a mistake :-) We can always commit a better fix later. @kbr- please still review this fix.